astropy/table/table.py

```diff
@@ -1795,12 +1755,58 @@ def add_column(self, col, index=None, name=None, rename_duplicate=False, copy=Tr
 
         To add several columns use add_columns.
         """
-        if index is None:
-            index = len(self.columns)
-        if name is not None:
-            name = (name,)
+        if default_name is None:
+            default_name = 'col{}'.format(len(self.columns))
+
+        # If value doesn't have a dtype and won't be added as a mixin then
+        # convert to a numpy array.
+        if not hasattr(col, 'dtype') and not self._is_mixin_for_table(col):
```

Comment:
Probably again more for follow-up, but in principle it would seem that the property that something should have is `shape` (and perhaps `info`). One could also think of `not isinstance(col, (np.ndarray, ShapedLikeNDArray))`.

(Maybe a follow-up issue on items like this - don't want to hold up this PR, which is mostly a refactor.)

Reply:
For this PR, might it make more sense to have this be part of `_convert_data_to_col`??

Reply:
Moved to #8943.

Reply:
> One could also think of not isinstance(col, (np.ndarray, ShapedLikeNDArray))

Currently there is no requirement that a mixin column inherit from `ShapedLikeNDArray`.  That is a very useful convenience, but not required.  (Though in practice I would not be surprised if some things break without that, and I just advised using `transpose()` in the cstack PR).

I thought one idea of duck arrays is to move away from type checking and more toward behavioral consistency.  Of course we are not there yet, but maybe for now this can be left until we change from mixin column to "duck column".  I like that name much better.

Reply:
Really all that's needed is indeed *being* reshapable/broadcastable and being able to deal with info. Anyway, the right place to discuss is the new issue!

Reply:
> Probably again more for follow-up, but in principle it would seem that the property that something should have is shape (and perhaps info). One could also think of not isinstance(col, (np.ndarray, ShapedLikeNDArray)).

Moved to #8944.